crates/profiling/src/vtune_linux.rs

```diff
@@ -0,0 +1,138 @@
+//! Adds support for profiling jitted code using VTune Amplifier
+//!
+//! Build:
+//!     cargo build --features=vtune
+//! Profile:
+//!     amplxe-cl -run-pass-thru=--no-altstack -v -collect hotspots target/debug/wasmtime --vtune test.wasm
+//!
+//! Note: amplxe-cl is a command-line tool for Vtune which should be installed.
+
+use anyhow::Result;
+use core::ptr;
+use ittapi_rs::*;
+use std::boxed::*;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::sync::Mutex;
+use wasmtime_environ::entity::PrimaryMap;
+use wasmtime_environ::wasm::DefinedFuncIndex;
+use wasmtime_environ::Module;
+use wasmtime_runtime::VMFunctionBody;
+
+/// Interface for driving the ittapi for VTune support
+pub struct VTuneAgent {
+    // Note that we use a mutex internally to serialize state updates
+    // since multiple threads may be sharing this agent.
+    state: Mutex<State>,
+}
+
+/// Interface for driving vtune
+#[derive(Clone, Debug, Default)]
+struct State {
+    /// Unique identifier for the jitted function
+    method_id: HashMap<(usize, DefinedFuncIndex), u32>,
+}
+
+impl VTuneAgent {
+    /// Intialize a VTuneAgent and write out the header
+    pub fn new() -> Result<Self> {
+        let state = State {
+            method_id: HashMap::new(),
+        };
+        Ok(VTuneAgent {
+            state: Mutex::new(state),
+        })
+    }
+}
+
+impl State {
+    /// Return the unique method ID for use with the ittapi
+    pub fn get_method_id(&mut self, module_id: usize, func_idx: DefinedFuncIndex) -> u32 {
+        let method_id;
+        unsafe {
+            method_id = self
+                .method_id
+                .entry((module_id, func_idx))
+                .or_insert_with(|| iJIT_GetNewMethodID());
+        }
+        *method_id
+    }
+
+    /// Load module
+    pub fn event_load(
+        &mut self,
+        method_id: u32,
+        module_name: &str,
+        method_name: &str,
+        addr: *const u8,
+        len: usize,
+    ) -> () {
+        let filename = "unknown_filename";
```

Comment:
Should this be an argument to `module_load` perhaps? To try to get the source filename if possible?

Reply:
er that is, an argument on the trait

Reply:
Sorry .. missed this comment earlier. Yeah, that would be great. I looked at instantiate_module and I don't see anywhere the path to the file is being preserved. Is this something possible now?

Reply:
Yeah we don't actually really ever have a filename right now, but it seems like it'd be best if the caller of the profiler would be inventing a filename like this rather than relying on each profiler to invent its own filename. We could eventually thread through a filename from something like `Module::from_file` to here when we get around to binding it.

Reply:
Ok, make sense. I'll add the argument and make up the name at the call site. 👍

Reply:
You know, I was thinking if filename should eventually come from the module, well the module is already passed in. We already pull the module_name (if exists) and method_name from the module and the only reason we don't pull the filename too is because it isn't there. I think ultimately we actually may not want this added to the interface. In the next push I've only added it to event load to see what you think, but if you still think it should be a part of the interface and passed in from the jit call site, then I can add it there too.